Switch to OpenBabel 3

rwest opened the pull request:
### Motivation or Problem
The release notes say OpenBabel 3 a major improvement
https://github.com/openbabel/openbabel/releases/tag/openbabel-3-0-0
but it does break some things in the API
https://open-babel.readthedocs.io/en/latest/UseTheLibrary/migration.html#migrating-to-3-0
so subsequent commits on this pull request should fix RMG to be compatible.

This pull request is currently a stub. Please feel free to contribute commits!

### Description of Changes
- update the environment.yaml to use the official updated cclib and openbabel conda packages.
- change import statements
- when reading radicals from OpenBabel molecules, calculate spin multiplicity based on undervalence of atoms, because OpenBabel no longer does this for you
- Mark a couple of broken unit tests as work_in_progress, because they seem to have malformed InChIs that we can no longer read, and I'm not sure what the correct molecule is supposed to be.

### Testing
- Tests run locally on my Mac.
- Continuous Integration tests will be run.

### Reviewer Tips
- if someone can figure out the broken InChIs in the unit tests, that'd be great.

Further detailed commentary is in the individual commit messages.

<!--
Checklist before submission:
 - Have you added appropriate unit tests?
 - Have you checked that all unit tests pass?
 - Is your code commented and understandable?
 - Have you updated related documentation?
 - Are the commits logically organized and informative?
 - Is your branch up to date with master?
-->

rwest commented:
Running tests locally I get only two problems now

```
======================================================================
ERROR: test_c6h6o4 (rmgpy.molecule.pathfinderTest.FindButadieneEndWithChargeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rwest/Code/RMG-Py/rmgpy/molecule/pathfinderTest.py", line 316, in test_c6h6o4
    mol = Molecule().from_inchi(inchi)
  File "rmgpy/molecule/molecule.py", line 1667, in rmgpy.molecule.molecule.Molecule.from_inchi
    def from_inchi(self, inchistr, backend='try-all', raise_atomtype_exception=True):
  File "rmgpy/molecule/molecule.py", line 1671, in rmgpy.molecule.molecule.Molecule.from_inchi
    translator.from_inchi(self, inchistr, backend, raise_atomtype_exception=raise_atomtype_exception)
  File "rmgpy/molecule/translator.py", line 254, in rmgpy.molecule.translator.from_inchi
    def from_inchi(mol, inchistr, backend='try-all', raise_atomtype_exception=True):
  File "rmgpy/molecule/translator.py", line 261, in rmgpy.molecule.translator.from_inchi
    return _read(mol, inchistr, 'inchi', backend, raise_atomtype_exception=raise_atomtype_exception)
  File "rmgpy/molecule/translator.py", line 503, in rmgpy.molecule.translator._read
    raise ValueError("Unable to correctly parse {0} with backend {1}.".format(identifier, backend))
ValueError: Unable to correctly parse InChI=1S/C6H6O4/c1-2-4-9-6(7)3-5-10-8/h2-3H,1,5H2 with backend try-all.

```

This is the same InChI that I had to mark as a work-in-progress because OpenBabel cannot parse it into a valid Molecule.
`InChI=1S/C6H6O4/c1-2-4-9-6(7)3-5-10-8/h2-3H,1,5H2`
Pasting it into https://rmg.mit.edu/molecule_search generates an adjacency list, but not a valid one (try clicking "Draw Molecule") so I'm not sure what it's meant to be. 

```
======================================================================
FAIL: test_nitrogenated_birad (rmgpy.molecule.pathfinderTest.FindAllylDelocalizationPathsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rwest/Code/RMG-Py/rmgpy/molecule/pathfinderTest.py", line 405, in test_nitrogenated_birad
    self.assertTrue(paths)
AssertionError: [] is not true
```

the test is 
```python
    def test_nitrogenated_birad(self):
        smiles = '[CH]=C[N]'
        mol = Molecule().from_smiles(smiles)
        paths = find_allyl_delocalization_paths(mol.atoms[3])
        self.assertTrue(paths)
```
after reading from_smiles the molecule is 
```
multiplicity 4
1 N u2 p1 c0 {2,S}
2 C u0 p0 c0 {1,S} {3,D} {4,S}
3 C u1 p0 c0 {2,D} {5,S}
4 H u0 p0 c0 {2,S}
5 H u0 p0 c0 {3,S}
```
compared to the RMG website which gives
```
multiplicity 4
1 C u1 p0 c0 {2,S} {3,D}
2 H u0 p0 c0 {1,S}
3 C u0 p0 c0 {1,D} {4,S} {5,S}
4 N u2 p1 c0 {3,S}
5 H u0 p0 c0 {3,S}
```
so I think it's just an atom ordering issue. I'll see if there's an alternative SMILES that passes both before and after.

rwest commented:
The unit tests now pass for me locally, on Mac OS X, where according to `conda list`  I am using 
`cclib                     1.7                pyhd3deb0d_0    conda-forge`
but on Travis it is preferring 
`cclib                     1.6.1.rmg        py37h39e3cac_0    rmg`
which uses the old openbabel imports.
The patch for the imports is at https://github.com/cclib/cclib/commit/3f73c94d1cb5c62fe5ba15af46e204fd2abab9da if someone wants to apply it to the "RMG" version of cclib and push new binaries to conda, but there may be other patches needed too, and it'd be vastly preferable to get back on track and use the official cclib.
Is there a reason we need our own? (and if so, why not for Mac OS X? or am I not testing Arkane or ARC locally or something?)

rwest commented:
See https://github.com/ReactionMechanismGenerator/RMG-Py/issues/756

rwest commented:
Current test failures in the travis log all to do with parsing mopac files with cclib:

```
======================================================================
8347ERROR: Test that Mocpac get_thermo_data() works correctly.
8348----------------------------------------------------------------------
8349Traceback (most recent call last):
8350  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mainTest.py", line 244, in test_get_thermo_data_mopac
8351    thermo1 = self.mop1.get_thermo_data(mol)
8352  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/main.py", line 216, in get_thermo_data
8353    thermo0 = qm_molecule_calculator.generate_thermo_data()
8354  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 359, in generate_thermo_data
8355    self.qm_data = self.generate_qm_data()
8356  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 309, in generate_qm_data
8357    success = self.run()
8358  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 149, in run
8359    return self.verify_output_file()
8360  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 215, in verify_output_file
8361    qm_data = self.parse()
8362  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 332, in parse
8363    parser = self.get_parser(self.output_file_path)
8364  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 230, in get_parser
8365    return cclib.parser.Mopac(output_file)
8366AttributeError: module 'cclib.parser' has no attribute 'Mopac'
8367
8368======================================================================
8369ERROR: Test that run_jobs() works properly.
8370----------------------------------------------------------------------
8371Traceback (most recent call last):
8372  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mainTest.py", line 304, in test_run_jobs
8373    qm.run_jobs(spc_list, procnum=1)
8374  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/main.py", line 253, in run_jobs
8375    _write_qm_files_star(qm_arg)
8376  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/main.py", line 264, in _write_qm_files_star
8377    return _write_qm_files(*args)
8378  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/main.py", line 271, in _write_qm_files
8379    quantum_mechanics.get_thermo_data(mol)
8380  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/main.py", line 216, in get_thermo_data
8381    thermo0 = qm_molecule_calculator.generate_thermo_data()
8382  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 359, in generate_thermo_data
8383    self.qm_data = self.generate_qm_data()
8384  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 309, in generate_qm_data
8385    success = self.run()
8386  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 149, in run
8387    return self.verify_output_file()
8388  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 215, in verify_output_file
8389    qm_data = self.parse()
8390  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 332, in parse
8391    parser = self.get_parser(self.output_file_path)
8392  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 230, in get_parser
8393    return cclib.parser.Mopac(output_file)
8394AttributeError: module 'cclib.parser' has no attribute 'Mopac'
8395
8396======================================================================
8397ERROR: Test that generate_thermo_data() works correctly for MOPAC PM3
8398----------------------------------------------------------------------
8399Traceback (most recent call last):
8400  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopacTest.py", line 88, in test_generate_thermo_data
8401    self.qmmol1.generate_thermo_data()
8402  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 359, in generate_thermo_data
8403    self.qm_data = self.generate_qm_data()
8404  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 309, in generate_qm_data
8405    success = self.run()
8406  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 149, in run
8407    return self.verify_output_file()
8408  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 215, in verify_output_file
8409    qm_data = self.parse()
8410  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 332, in parse
8411    parser = self.get_parser(self.output_file_path)
8412  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 230, in get_parser
8413    return cclib.parser.Mopac(output_file)
8414AttributeError: module 'cclib.parser' has no attribute 'Mopac'
8415
8416======================================================================
8417ERROR: Test that generate_thermo_data() can load thermo from the previous MOPAC PM3 run.
8418----------------------------------------------------------------------
8419Traceback (most recent call last):
8420  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopacTest.py", line 107, in test_load_thermo_data
8421    self.qmmol1.generate_thermo_data()
8422  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 359, in generate_thermo_data
8423    self.qm_data = self.generate_qm_data()
8424  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 298, in generate_qm_data
8425    if self.verify_output_file():
8426  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 215, in verify_output_file
8427    qm_data = self.parse()
8428  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 332, in parse
8429    parser = self.get_parser(self.output_file_path)
8430  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 230, in get_parser
8431    return cclib.parser.Mopac(output_file)
8432AttributeError: module 'cclib.parser' has no attribute 'Mopac'
8433
8434======================================================================
8435ERROR: Test that generate_thermo_data() works correctly for MOPAC PM6
8436----------------------------------------------------------------------
8437Traceback (most recent call last):
8438  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopacTest.py", line 152, in test_generate_thermo_data
8439    self.qmmol1.generate_thermo_data()
8440  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 359, in generate_thermo_data
8441    self.qm_data = self.generate_qm_data()
8442  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 309, in generate_qm_data
8443    success = self.run()
8444  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 149, in run
8445    return self.verify_output_file()
8446  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 215, in verify_output_file
8447    qm_data = self.parse()
8448  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 332, in parse
8449    parser = self.get_parser(self.output_file_path)
8450  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 230, in get_parser
8451    return cclib.parser.Mopac(output_file)
8452AttributeError: module 'cclib.parser' has no attribute 'Mopac'
8453
8454======================================================================
8455ERROR: Test that generate_thermo_data() can load thermo from the previous MOPAC PM6 run.
8456----------------------------------------------------------------------
8457Traceback (most recent call last):
8458  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopacTest.py", line 172, in test_load_thermo_data
8459    self.qmmol1.generate_thermo_data()
8460  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 359, in generate_thermo_data
8461    self.qm_data = self.generate_qm_data()
8462  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 298, in generate_qm_data
8463    if self.verify_output_file():
8464  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 215, in verify_output_file
8465    qm_data = self.parse()
8466  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 332, in parse
8467    parser = self.get_parser(self.output_file_path)
8468  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 230, in get_parser
8469    return cclib.parser.Mopac(output_file)
8470AttributeError: module 'cclib.parser' has no attribute 'Mopac'
8471
8472======================================================================
8473ERROR: Test that generate_thermo_data() works correctly for MOPAC PM7
8474----------------------------------------------------------------------
8475Traceback (most recent call last):
8476  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopacTest.py", line 218, in test_generate_thermo_data
8477    self.qmmol1.generate_thermo_data()
8478  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 359, in generate_thermo_data
8479    self.qm_data = self.generate_qm_data()
8480  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 309, in generate_qm_data
8481    success = self.run()
8482  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 149, in run
8483    return self.verify_output_file()
8484  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 215, in verify_output_file
8485    qm_data = self.parse()
8486  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 332, in parse
8487    parser = self.get_parser(self.output_file_path)
8488  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 230, in get_parser
8489    return cclib.parser.Mopac(output_file)
8490AttributeError: module 'cclib.parser' has no attribute 'Mopac'
8491
8492======================================================================
8493ERROR: Test that generate_thermo_data() can load thermo from the previous MOPAC PM7 run.
8494----------------------------------------------------------------------
8495Traceback (most recent call last):
8496  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopacTest.py", line 238, in test_load_thermo_data
8497    self.qmmol1.generate_thermo_data()
8498  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 359, in generate_thermo_data
8499    self.qm_data = self.generate_qm_data()
8500  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 298, in generate_qm_data
8501    if self.verify_output_file():
8502  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 215, in verify_output_file
8503    qm_data = self.parse()
8504  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/molecule.py", line 332, in parse
8505    parser = self.get_parser(self.output_file_path)
8506  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/qm/mopac.py", line 230, in get_parser
8507    return cclib.parser.Mopac(output_file)
8508AttributeError: module 'cclib.parser' has no attribute 'Mopac'
```

rwest commented on rmgpy/molecule/pathfinderTest.py:
```diff
@@ -312,18 +312,14 @@ def test_c5h6o2(self):
-        # Can't read this InChI:
-        # the old OpenBabel could read it, but the resulting molecule had an invalid adjacency_list
-        inchi = "InChI=1S/C6H6O4/c1-2-4-9-6(7)3-5-10-8/h2-3H,1,5H2"
+        inchi = "InChI=1S/C8H14O4S/c1-3-6-13(2,11)7-8(9)4-5-12-10/h3,6H,1,4-5,7H2,2H3,(H-,10,11)"
```
One suggestion, can we change the name of the test? (since it's no longer c6h6o4) and perhaps add a docstring or comment so future people know what we're trying to test (which would have helped previously).

rwest replied:
I amended the commit when rebasing and changed the name of the test.

mjohnson541 commented on rmgpy/molecule/pathfinderTest.py:
```diff
@@ -28,6 +28,7 @@
 ###############################################################################
 
 import unittest
+from external.wip import work_in_progress
```
Can we clean up the git history here so we don't have unnecessarily overlapping changes?

rwest replied:
done.

mjohnson541 reviewed (commented):
This looks good to me, just the changes to pathfinderTest test_c6h6o4 should probably be squashed.

alongd commented:
Looks good to me as well. I guess rmg-tests fail due to known reasons?
I'll create a twin ob3 PR in ARC.

alongd commented:
I get a wrong Molecule object for phenanthrene on this branch:
![image](https://user-images.githubusercontent.com/16158262/115147254-d96d7300-a062-11eb-87a4-d3b11bf5a4b6.png)

rwest commented:
Oh :-(

http://openbabel.org/docs/current/Features/Radicals.html

**SMILES extensions for radicals**

Finishes with:

> This extension is not as robust or as carefully considered as standard SMILES and should be used with restraint. A structure that uses c as a radical centre close to aromatic carbons can be confusing to read, and Open Babel’s SMILES parser can also be confused. For example, it recognizes c1ccccc1c as the benzyl radical, but it doesn’t like c1cc(c)ccc1. Radical centres should not be involved in ring closure: for cyclohexyl radical C1cCCCC1 is ok, but c1CCCCC1 is not.

The URL says it's "current" documentation but the HTML says 2.3.1

Guess one could post to the openbabel user group mailing list

rwest commented:
It might be our conversion from OBMol to Molecule. Or it might be OpenBabel's parsing of the SMILES.

alongd commented:
I get the correct Molecule, though, when I use the full SMILES `C1=CCC2=C3CC=CCC3=CC=C2C1` w/o the lowercase c's

alongd commented:
I requested "posting permission" for the open babel forum.

I got the same problem with `c1cc1O`. Other `ccc..` SMILES I tried were OK. Meanwhile, we could direct users to avoid using this syntax, or wait for the bug to be resolved? Both aren't great options

alongd commented:
Citing from @rwest's link:

_This extension is not as robust or as carefully considered as standard SMILES and should be used with restraint. A structure that uses c as a radical centre close to aromatic carbons can be confusing to read, and Open Babel’s SMILES parser can also be confused_

So I guess it's OK if it's "designed/known", although the previous version supported these cases better. We might want to document this though?

mjohnson541 commented:
Do we want to wait and not include this in the release?

rwest commented:
Does `c1ccc2c(c1)ccc3ccccc23` work?

Not sure what the solution is. 

Issue a warning when using lower case letters in a SMILES that things may not be what you expect? Add a note in documentation?

rwest commented:
What were you expecting for `c1cc1O` ? The current rmg.mit.edu website gives cyclopropanol, as does `OC1CC1`, and https://cactus.nci.nih.gov/chemical/structure/c1cc1O/names/xml

alongd commented:
For `c1cc1O` I do expect to get `OC1CC1`, but on this branch I get:
```
multiplicity 2
1 O u0 p2 c0 {2,S} {7,S}
2 C u0 p0 c0 {1,S} {3,S} {4,D}
3 C u1 p0 c0 {2,S} {4,S} {6,S}
4 C u0 p0 c0 {2,D} {3,S} {5,S}
5 H u0 p0 c0 {4,S}
6 H u0 p0 c0 {3,S}
7 H u0 p0 c0 {1,S}
```
![image](https://user-images.githubusercontent.com/16158262/115509313-f2ef0480-a286-11eb-8f50-fe2c53732158.png)

for `c1ccc2c(c1)ccc3ccccc23` I do get a reasonable solution on this branch:
![image](https://user-images.githubusercontent.com/16158262/115509460-20d44900-a287-11eb-8bb2-1f5cb4628c8c.png)

I think this is not an ordinary "bug", but rather perhaps a "design choice" by OB's devs. I recommend documenting it. We could make RMG crush if any `c1cc....` sort of SMILES shortcut ends up as a radical. You can certainty aim to have radicals with `c1ccc...`'s but it is likely that you got the SMILES correctly if you end up with a multiplicity 1, but I haven't tested thoroughly. This is a bit of a headache and not hermetic. Perhaps don't allow lowercase SMILES will be a better option?

alongd commented:
You are absolutely correct. Thanks for getting to the bottom of this, I agree with the interpretation that NOW OB behaves "as expected". I'm still uncomfortable with the interpretation of the (weird) `c1=ccc2c3cc=ccc3ccc2c1` structure. But I see you point re "c equals CH", and that an explicit double bond is indeed ambiguous. Please go ahead with getting this PR in, as far as I can see it looks good. Thanks!

mjohnson541 commented:
Alright lets get this in. Can you rebase?

rwest commented:
Rebased, and added a small note to the documentation.

Out of curiosity, I extended the notebook to compare the RDKit back end. In many cases it does similar things, it ends up with similar isomers for the C14H6 case but with different electronic structure. And it indeed crashes for some of the non-aromatic cases.

[NewOpenBabel.pdf](https://github.com/ReactionMechanismGenerator/RMG-Py/files/6354544/NewOpenBabel.pdf)
[NewOpenBabel.ipynb.zip](https://github.com/ReactionMechanismGenerator/RMG-Py/files/6354546/NewOpenBabel.ipynb.zip)
